akka-stream/src/main/scala/akka/stream/impl/TraversalBuilder.scala

```diff
@@ -684,8 +684,11 @@ final case class LinearTraversalBuilder(
    */
   def append[A, B, C](toAppend: LinearTraversalBuilder, matCompose: (A, B) ⇒ C): LinearTraversalBuilder = {
 
-    if (toAppend.isEmpty) this
-    else if (this.isEmpty) {
+    if (toAppend.isEmpty) {
+      copy(
+        traversalSoFar = PushNotUsed.concat(LinearTraversalBuilder.addMatCompose(traversalSoFar, matCompose))
```

Comment:
the "traversal" of the empty linear is `PushNotUsed` as only it does is add a new materialized value of `NotUsed`. The `concat` call will actually remove this if immediately followed by a `Pop`.